src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs

```diff
@@ -923,36 +926,49 @@ private Type FindType(int typeIndex)
             }
             if (_typeTable[typeIndex] == null)
             {
-                long oldPos = _store.BaseStream.Position;
-                try
-                {
-                    _store.BaseStream.Position = _typeNamePositions[typeIndex];
-                    string typeName = _store.ReadString();
-                    _typeTable[typeIndex] = Type.GetType(typeName, true);
-                }
-                // If serialization isn't supported, we convert FileNotFoundException to
-                // NotSupportedException for consistency with v2. This is a corner-case, but the
-                // idea is that we want to give the user a more accurate error message. Even if
-                // the dependency were found, we know it will require serialization since it
-                // can't be one of the types we special case. So if the dependency were found,
-                // it would go down the serialization code path, resulting in NotSupported for
-                // SKUs without serialization.
-                //
-                // We don't want to regress the expected case by checking the type info before
-                // getting to Type.GetType -- this is costly with v1 resource formats.
-                catch (FileNotFoundException)
+                if (AllowReflectionForNonPrimitiveTypes)
                 {
-                    throw new NotSupportedException(SR.NotSupported_ResourceObjectSerialization);
+                    UseReflectionToGetType(typeIndex);
                 }
-                finally
+                else
                 {
-                    _store.BaseStream.Position = oldPos;
+                    throw new NotSupportedException(SR.ResourceManager_ReflectionNotAllowed);
                 }
             }
             Debug.Assert(_typeTable[typeIndex] != null, "Should have found a type!");
```

Comment:
While I agree with you and it could get moved inside `UseReflectionToGetType`, I'd rather keep it here since in the next line we are using `!` to assert that the returned value shouldn't be null, so I would prefer to keep the assert right next to the return statement with the nullable annotation. If you still have a strong opinion to have it moved, I can do that.